### PR TITLE
feat(github-action): Support updating digest pinned actions

### LINKED ISSUE #7537: Pin GitHub Action Git SHA (Digest Pinning)
**What would you like Renovate to be able to do?**

Pin a GitHub action to a specific Git SHA (digest pinning).

**Did you already have any implementation ideas?**

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

**Is this a feature you'd be interested in implementing yourself?**

**References**

See #7516 for related discussion.

### PR DESCRIPTION
**Pull request by viceice**
## Changes:

Support updating github action sha pinned actions.

Sample pr's:
- https://github.com/viceice-tests/gh-action-digests/pull/2
- https://github.com/viceice-tests/gh-action-digests/pull/3
- https://github.com/viceice-tests/gh-action-digests/pull/4
- https://github.com/viceice-tests/gh-action-digests/pull/5
- https://github.com/viceice-tests/gh-action-digests/pull/6
- https://github.com/viceice-tests/gh-action-digests/pull/7

## Context:

closes #7537

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — APPROVED** [review]

**Comment by rarkins** [comment]
Can you also create feature issues for:
- supporting an informative "version" field for Renovate to use as currentValue, so that we don't need to propose a PR every time there's a commit to the action's main branch?
- supporting displaying a gitRef deep link diff (could apply to more than just actions)

I think we need both of these in order for actions digest PRs to be more usable

**Review by rarkins — DISMISSED** [review]

**Comment by viceice** [comment]
add tag follow support for pinned actions via renovate comment

**Comment by rarkins** [comment]
Do we support pinning digests for actions? And can it move the tag to a comment if it does? :)

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/manager/github-actions/extract.ts [hunk 1]
```diff
@@ -5,6 +5,11 @@ import * as dockerVersioning from '../../versioning/docker';
 import { getDep } from '../dockerfile/extract';
 import type { PackageDependency, PackageFile } from '../types';
 
+const dockerRe = /^\s+uses: docker:\/\/([^"]+)\s*$/;
+const actionRe =
+  /^\s+-?\s+?uses: (?<depName>[\w-]+\/[\w-]+)(?<path>.*)?@(?<currentValue>.+?)(?: # renovate: tag=(?<tag>.+?))?\s*?$/;
+const sha256Re = /^[a-z0-9]{40}$/;
```
[the remark this answers is not in the document]
Comment (viceice) [reply lib/manager/github-actions/extract.ts]: Now supports SHA1 and SHA256 to be future proof

**Comment by viceice** [comment]
> Do we support pinning digests for actions? And can it move the tag to a comment if it does? :)

Not yet, working on it 🙃

File: lib/manager/github-actions/readme.md [hunk 2]
```diff
@@ -1 +1,17 @@
...
+on: [push]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@af513c7a016048ae468971c52ed77d9562c7c819 # renovate: tag=v1.0.0
+```
+
+Renovate will update the commit SHA1 but follow the GitHub tag you specified.
```
Comment (HonkingGoose) [inline lib/manager/github-actions/readme.md:17]:
```suggestion
Renovate will update the commit SHA but follow the GitHub tag you specified.
Renovate can update digests that use SHA1 and SHA256 algorithms.
```

Use generic SHA name, and explain that Renovate can update SHA1 and SHA256 digest hashes.

**Review by HonkingGoose — COMMENTED** [review]
Some docs changes to keep up with new feature.

**Comment by viceice** [comment]
It now also supports pin digests with auto replace 🎉

File: lib/workers/branch/auto-replace.ts [hunk 3]
```diff
@@ -22,7 +22,7 @@ export async function confirmIfDepUpdated(
     pinDigests,
   } = upgrade;
   const extractPackageFile = get(manager, 'extractPackageFile');
-  let newUpgrade;
+  let newUpgrade: PackageDependency;
```
Comment (viceice) [inline lib/workers/branch/auto-replace.ts:25]: found while debugging

**Review by viceice — COMMENTED** [review]
Anything missing?

File: lib/manager/github-actions/readme.md [hunk 4]
```diff
@@ -1 +1,26 @@
...
+Renovate will update the commit SHA but follow the GitHub tag you specified.
+Renovate can update digests that use SHA1 and SHA256 algorithms.
+You can also use the following PackageRule to automatically pin action digests.
+
+```json
+{
+  "matchDepTypes": ["action"],
+  "pinDigests": true
+}
+```
```
Comment (HonkingGoose) [inline lib/manager/github-actions/readme.md:26]: How about we add a new config preset? I'm thinking of something like `:automaticallyPinGitHubActionDigests`. Or a suitable short name. 😉

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
Thought I already commented here but can't see it. Need to test then use case where tag version remains same but digest is updated. Valid tag with bogus sha in base branch should reproduce it. 

Also I noted that the pr body comment for pinning is wrong/outdated but it's unrelated to this PR

**Comment by viceice** [comment]
> Thought I already commented here but can't see it. Need to test the use case where tag version remains same but digest is updated. Valid tag with bogus sha in base branch should reproduce it.
> 
> Also I noted that the pr body comment for pinning is wrong/outdated but it's unrelated to this PR

Sure, creating sample pr

**Comment by viceice** [comment]
> Thought I already commented here but can't see it. Need to test the use case where tag version remains same but digest is updated. Valid tag with bogus sha in base branch should reproduce it.
> 
> Also I noted that the pr body comment for pinning is wrong/outdated but it's unrelated to this PR

@rarkins see https://github.com/viceice-tests/gh-action-digests/pull/7 works 🎉

File: lib/manager/github-actions/readme.md (hunk 4 above)
Comment (viceice) [reply lib/manager/github-actions/readme.md:27]: done https://github.com/renovatebot/renovate/pull/10835/commits/f932ba09a936566b9d5ad78aadd963c7aa13f894

**Review by rarkins — DISMISSED** [review]

File: lib/manager/github-actions/readme.md [hunk 5]
```diff
@@ -1 +1,27 @@
...
+Renovate can update digests that use SHA1 and SHA256 algorithms.
+
+If you want to automatically pin action digests put the following configuration in a new `packageRules` array :
+
+```json
+{
+  "matchDepTypes": ["action"],
+  "pinDigests": true
+}
+```
```
Comment (viceice) [inline lib/manager/github-actions/readme.md:27]:
```suggestion
If you want to automatically pin action digests put the following configuration in the extends array : 

```json
{
  "extends": ["helper:pinGitHubActionDigests"]
}
```
```

What about this

**Review by rarkins — APPROVED** [review]

**Pull request merged**